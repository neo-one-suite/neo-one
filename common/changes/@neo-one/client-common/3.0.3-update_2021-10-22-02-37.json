{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Update nef model and types",
      "type": "minor"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "29364457+spencercorwin@users.noreply.github.com"
}
