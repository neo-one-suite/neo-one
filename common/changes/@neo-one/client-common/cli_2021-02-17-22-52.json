{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Fix minor bugs.",
      "type": "patch"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "spencercorwin@icloud.com"
}
