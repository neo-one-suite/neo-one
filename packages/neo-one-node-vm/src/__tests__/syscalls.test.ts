import { common, ScriptBuilder, TriggerType } from '@neo-one/client-common';
import { Block, ConsensusData, Witness } from '@neo-one/node-core';
import { BN } from 'bn.js';
import _ from 'lodash';
import { ApplicationEngine } from '../ApplicationEngine';
import { Dispatcher } from '../Dispatcher';

describe('Application Engine SysCall Tests', () => {
  const dispatcher = new Dispatcher();
  const engine = new ApplicationEngine(dispatcher);
  let script: ScriptBuilder;
  beforeEach(() => {
    dispatcher.reset();
    engine.create({
      trigger: TriggerType.Application,
      gas: 0,
      testMode: true,
    });

    script = new ScriptBuilder();
  });

  describe('System.Json.Deserialize', () => {
    test('Halt -- Good values', () => {
      script.emitPushString('123');
      script.emitSysCall('System.Json.Deserialize');
      script.emitPushString('null');
      script.emitSysCall('System.Json.Deserialize');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('HALT');

      const resultStack = engine.resultStack;
      expect(resultStack[0].isNull).toEqual(true);
      expect(resultStack[1].getInteger().toNumber()).toEqual(123);
    });

    test('Fault -- Wrong Json', () => {
      script.emitPushString('***');
      script.emitSysCall('System.Json.Deserialize');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('FAULT');
      expect(engine.resultStack.length).toEqual(0);
    });

    test('Fault -- No decimals', () => {
      script.emitPushString('123.45');
      script.emitSysCall('System.Json.Deserialize');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('FAULT');
      expect(engine.resultStack.length).toEqual(0);
    });
  });

  describe('System.Json.Serialize', () => {
    test('Halt -- Good values', () => {
      script.emitPushInt(5);
      script.emitSysCall('System.Json.Serialize');
      script.emitOp('PUSH0');
      script.emitOp('NOT');
      script.emitSysCall('System.Json.Serialize');
      script.emitPushString('test');
      script.emitSysCall('System.Json.Serialize');
      script.emitOp('PUSHNULL');
      script.emitSysCall('System.Json.Serialize');
      script.emitOp('NEWMAP');
      script.emitOp('DUP');
      script.emitPushString('key');
      script.emitPushString('value');
      script.emitOp('SETITEM');
      script.emitSysCall('System.Json.Serialize');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('HALT');

      const resultStack = engine.resultStack;
      expect(resultStack.length).toEqual(5);

      expect(resultStack[0].getString()).toEqual('{"key":"value"}');
      expect(resultStack[1].getString()).toEqual('null');
      expect(resultStack[2].getString()).toEqual('"test"');
      expect(resultStack[3].getString()).toEqual('true');
      expect(resultStack[4].getString()).toEqual('5');
    });

    test('Fault -- Bad values', () => {
      script.emitSysCall('System.Storage.GetContext');
      script.emitSysCall('System.Json.Serialize');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('FAULT');
      expect(engine.resultStack.length).toEqual(0);
    });
  });

  test('System.Callback.Invoke -- Halt', () => {
    engine.create({
      trigger: TriggerType.Application,
      gas: 1,
      testMode: false,
    });

    script = new ScriptBuilder();

    script.emitPushInt(5); // Callback argument 1
    script.emitPushInt(1); // Callback argument 2
    script.emitPushInt(2); // ParamCount
    script.emitOp('PACK');
    script.emitPushInt(2); // ParamCount
    script.emitOp('PUSHA', new BN(200).toArrayLike(Buffer, 'le', 4)); // -> Nop area
    script.emitSysCall('System.Callback.Create');
    script.emitSysCall('System.Callback.Invoke');
    script.emitOp('RET');
    _.range(250).forEach(() => {
      script.emitOp('NOP');
    });
    script.emitOp('SUB');
    script.emitOp('RET');

    engine.loadScript(script.build());
    const state = engine.execute();
    expect(state).toEqual('HALT');

    const resultStack = engine.resultStack;
    expect(resultStack.length).toEqual(1);
    expect(resultStack[0].getInteger().eq(new BN(4))).toEqual(true);
  });

  test('System.Callback.CreateFromSyscall -- Halt', () => {
    engine.create({
      trigger: TriggerType.Application,
      gas: 1,
      testMode: false,
    });

    script = new ScriptBuilder();

    script.emitPush(Buffer.from([]));
    script.emitPushInt(1);
    script.emitOp('PACK');
    script.emitSysCallName('Neo.Crypto.SHA256');
    script.emitSysCall('System.Callback.CreateFromSyscall');
    script.emitSysCall('System.Callback.Invoke');

    engine.loadScript(script.build());
    const state = engine.execute();
    expect(state).toEqual('HALT');

    const resultStack = engine.resultStack;
    expect(resultStack.length).toEqual(1);
    expect(resultStack[0].getBuffer().toString('hex')).toEqual(
      'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
    );
  });

  describe('System.Blockchain.GetBlock', () => {
    const block = new Block({
      index: 0,
      timestamp: new BN(2),
      version: 3,
      witness: new Witness({ invocation: Buffer.from([]), verification: Buffer.from([]) }),
      previousHash: common.ZERO_UINT256,
      merkleRoot: common.ZERO_UINT256,
      nextConsensus: common.ZERO_UINT160,
      consensusData: new ConsensusData({ nonce: new BN(1), primaryIndex: 1 }),
      transactions: [],
    });

    test('With Snapshot option -- Halt', () => {
      engine.create({
        trigger: TriggerType.Application,
        gas: 0,
        testMode: true,
        snapshot: 'main',
      });
      script.emitPushUInt256(block.hash);
      script.emitSysCall('System.Blockchain.GetBlock');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('HALT');

      const stack = engine.resultStack;
      expect(stack.length).toEqual(1);
      expect(stack[0].isNull).toEqual(true);
    });

    test('Without Snapshot option -- Fault', () => {
      engine.create({ trigger: TriggerType.Application, gas: 0, testMode: true });
      script.emitPushUInt256(block.hash);
      script.emitSysCall('System.Blockchain.GetBlock');

      engine.loadScript(script.build());
      const state = engine.execute();
      expect(state).toEqual('FAULT');
    });
  });
});
