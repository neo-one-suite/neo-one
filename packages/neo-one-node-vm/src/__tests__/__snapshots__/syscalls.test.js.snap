// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syscalls Neo.Account.GetBalance 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Account.GetBalance 2`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Account.GetBalance 3`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Account.GetBalance 4`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Account.GetScriptHash 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Account.GetVotes 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetAdmin 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetAmount 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetAssetId 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetAssetType 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetAvailable 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetIssuer 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetOwner 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Asset.GetPrecision 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Block.GetTransaction 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Block.GetTransactionCount 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Block.GetTransactions 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetAccount 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetAsset 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetBlock 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetContract 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetHeader 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Blockchain.GetTransaction 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Contract.GetScript 1`] = `
Array [
  Array [
    Object {
      "hash": "3775292229eccdf904f16fff8e83e7cffdc0f0ce",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetConsensusData 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetHash 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetIndex 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetMerkleRoot 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetNextConsensus 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetPrevHash 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetTimestamp 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Header.GetVersion 1`] = `
Array [
  Array [
    Object {
      "hashOrIndex": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.InvocationTransaction.GetScript 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Iterator.Next 1`] = `
Array [
  Array [
    Object {
      "hash": "686302f409a2f3a10c2a7e547703d5032b877c75",
    },
  ],
]
`;

exports[`syscalls Neo.Iterator.Next 2`] = `
Array [
  Array [
    Object {
      "hash": "686302f409a2f3a10c2a7e547703d5032b877c75",
      "prefix": "01",
    },
  ],
]
`;

exports[`syscalls Neo.Iterator.Next 3`] = `
Array [
  Array [
    Object {
      "hash": "686302f409a2f3a10c2a7e547703d5032b877c75",
    },
  ],
]
`;

exports[`syscalls Neo.Iterator.Next 4`] = `
Array [
  Array [
    Object {
      "hash": "686302f409a2f3a10c2a7e547703d5032b877c75",
      "prefix": "01",
    },
  ],
]
`;

exports[`syscalls Neo.Runtime.Log 1`] = `
Array [
  Array [
    LogAction {
      "blockHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "blockIndex": -1,
      "index": 0,
      "message": "foo",
      "scriptHash": "720f6e0afa42c0ba6b0a7a414ef94b10ac966c78",
      "serializeWire": [Function],
      "transactionHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "transactionIndex": -1,
      "type": 0,
      "version": 0,
    },
  ],
]
`;

exports[`syscalls Neo.Runtime.Notify 1`] = `
Array [
  Array [
    NotificationAction {
      "args": Array [
        Object {
          "__$type": "IntegerContractParameter",
          "type": 2,
          "value": "1",
        },
      ],
      "blockHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "blockIndex": -1,
      "index": 0,
      "scriptHash": "20bdad340ffefc952c253989bb24ca0de4857a36",
      "serializeWire": [Function],
      "transactionHash": "0000000000000000000000000000000000000000000000000000000000000000",
      "transactionIndex": -1,
      "type": 1,
      "version": 0,
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Delete 1`] = `
Array [
  Array [
    Object {
      "hash": "bff26630a34d202045d1fc72a9f905b0f308f43d",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Delete 2`] = `
Array [
  Array [
    Object {
      "hash": "bff26630a34d202045d1fc72a9f905b0f308f43d",
      "key": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Find 1`] = `[Function]`;

exports[`syscalls Neo.Storage.Find 2`] = `
Array [
  Array [
    Object {
      "hash": "172a85ad477e8853f60c2bdce6e92b5492c4b5b3",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Find 3`] = `
Array [
  Array [
    Object {
      "hash": "172a85ad477e8853f60c2bdce6e92b5492c4b5b3",
      "prefix": "01",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Get 1`] = `
Array [
  Array [
    Object {
      "hash": "172a85ad477e8853f60c2bdce6e92b5492c4b5b3",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Get 2`] = `
Array [
  Array [
    Object {
      "hash": "172a85ad477e8853f60c2bdce6e92b5492c4b5b3",
      "key": "01",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 1`] = `
Array [
  Array [
    Object {
      "hash": "5a4dc5955ca2dcef435f4cdc40439ebf49632abd",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 2`] = `
Array [
  Array [
    Object {
      "hash": "5a4dc5955ca2dcef435f4cdc40439ebf49632abd",
      "key": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 3`] = `
Array [
  Array [
    StorageItem {
      "__size": [Function],
      "equals": [Function],
      "hash": "5a4dc5955ca2dcef435f4cdc40439ebf49632abd",
      "key": "",
      "serializeWire": [Function],
      "value": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 4`] = `
Array [
  Array [
    Object {
      "hash": "65b899fa49b9c7c16211de54af874319eb36ada3",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 5`] = `
Array [
  Array [
    Object {
      "hash": "65b899fa49b9c7c16211de54af874319eb36ada3",
      "key": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 6`] = `
Array [
  Array [
    StorageItem {
      "__size": [Function],
      "equals": [Function],
      "hash": "65b899fa49b9c7c16211de54af874319eb36ada3",
      "key": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "serializeWire": [Function],
      "value": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 7`] = `
Array [
  Array [
    Object {
      "hash": "c713aebd42a0dc3fe812df408fcaf1b50a6dadb3",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 8`] = `
Array [
  Array [
    Object {
      "hash": "c713aebd42a0dc3fe812df408fcaf1b50a6dadb3",
      "key": "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 9`] = `
Array [
  Array [
    StorageItem {
      "__size": [Function],
      "equals": [Function],
      "hash": "c713aebd42a0dc3fe812df408fcaf1b50a6dadb3",
      "key": "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "serializeWire": [Function],
      "value": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 10`] = `
Array [
  Array [
    Object {
      "hash": "6b0c054a07b8264a8347d9cdd1d49a6efb6631ef",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 11`] = `
Array [
  Array [
    Object {
      "hash": "6b0c054a07b8264a8347d9cdd1d49a6efb6631ef",
      "key": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 12`] = `
Array [
  Array [
    StorageItem {
      "__size": [Function],
      "equals": [Function],
      "hash": "6b0c054a07b8264a8347d9cdd1d49a6efb6631ef",
      "key": "",
      "serializeWire": [Function],
      "value": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 13`] = `
Array [
  Array [
    Object {
      "hash": "cf131c8ec9c75f49d824958f33df72b8e18980b5",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 14`] = `
Array [
  Array [
    Object {
      "hash": "cf131c8ec9c75f49d824958f33df72b8e18980b5",
      "key": "",
    },
  ],
]
`;

exports[`syscalls Neo.Storage.Put 15`] = `
Array [
  Array [
    StorageItem {
      "__size": [Function],
      "equals": [Function],
      "hash": "cf131c8ec9c75f49d824958f33df72b8e18980b5",
      "key": "",
      "serializeWire": [Function],
      "value": "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetAttributes 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetHash 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetInputs 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetOutputs 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetReferences 1`] = `
Array [
  Array [
    Object {
      "__$type": "Input",
      "hash": "0000000000000000000000000000000000000000000000000000000000000000",
      "index": 0,
      "size": 34,
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetReferences 2`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetType 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetUnspentCoins 1`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetUnspentCoins 2`] = `
Array [
  Array [
    Object {
      "hash": "df09c9981eb0c4a2a182ce6090148647a2f4f6053c970380a645ed63b91ce258",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetUnspentCoins 3`] = `
Array [
  Array [
    Object {
      "hash": "0303030303030303030303030303030303030303030303030303030303030303",
    },
  ],
]
`;

exports[`syscalls Neo.Transaction.GetUnspentCoins 4`] = `
Array [
  Array [
    Object {
      "hash": "df09c9981eb0c4a2a182ce6090148647a2f4f6053c970380a645ed63b91ce258",
    },
  ],
]
`;
